Payjoin (#993)
BitcoinDesign/Guide: guide/case-studies/payjoin.md

- *[Nicolas Dorier](https://twitter.com/nicolasdorier) for reviewing the case study content and engaging on payjoin in general*
- *Finally, thanks [Dan Gould](https://twitter.com/bitgould) for invaluable guidance, patience and passion*

## Resources

- [BIP-78](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki)
- [Bitcoin Optech topics: Payjoin](https://bitcoinops.org/en/topics/payjoin)
- [Payjoin Process Flows](https://www.figma.com/file/NzMvwyzP7x5jfGmwNUKRov/PayJoin-Process-Flows?node-id=0%3A1&t=wSwewTOkQddWhagl-1)
- [Payjoin User Flows](https://www.figma.com/file/69uUDWVc8N9t5Bej8pZEsF/PayJoin-User-Flows?node-id=0%3A1&t=8F4jOa71i6X1Slbz-1)
- [Investigation: Petting Payjoins in the Wild](https://docs.google.com/document/d/1_de2pkMREGpZQwOefQdH-MZNSUPazObgRK1cruiWBmo/edit?usp=sharing)
- [BTCPay Server Payjoin Guide](https://docs.btcpayserver.org/Payjoin/#btcpay-server-payjoin-guide)

Next, view the [resources]({{ '/guide/resources' | relative_url }}) that can help you build better bitcoin products.

{% include next-previous.html
   previousUrl = "/guide/case-studies/blixt-wallet"

>    previousName = "Case studies"

Reviewer: ```suggestion
   previousName = "Blixt wallet case study"
```

---
layout: guide
title: Payments with Payjoin
description: <>
nav_order: 5
parent: Case studies
permalink: /guide/case-studies/payjoin-case-study/
main_classes: -no-top-padding
image: https://bitcoin.design/assets/images/guide/case-studies/blixt-wallet/blixt-wallet-preview.jpg
---

<!--

Editor's notes

Illustration sources: <>

-->

<!-----
Conversion notes:

* Docs to Markdown version 1.0β34
* Sun Mar 05 2023 13:50:07 GMT-0800 (PST)
* Source doc: [restructure] Payments with Payjoin: A Case Study
* Tables are currently converted to HTML tables.
* This document has images: check for >>>>>  gd2md-html alert:  inline image link in generated source and store images to your server. NOTE: Images in exported zip file from Google Docs may not appear in  the same order as they do in your doc. Please check the images!

----->

# Payments with Payjoin: A Case Study
{:.no_toc}

---

<div class="glossary-toc" markdown="1">
* Table of contents
{:toc}
</div>

---

## Introduction

> Payjoin transactions can be a powerful tool providing privacy, fee-savings and UTXO consolidation benefits to users. They also present challenges and trade-offs due to their synchronous, multi-party nature.

Reviewer 2: ```suggestion
Payjoin transactions can be a powerful tool providing privacy, fee-savings and UTXO consolidation benefits to users. They also present challenges and trade-offs due to their interactive nature.
```

I think all but self-spend transactions  are multi-party. They key with payjoin is that  it is *interactive*. Synchronicity may be considered a detail.

- Proposes specific user flows and where possible, working designs for the same.

The high-level findings from this case study are that while the setup is more involved for the receiver, little-to-no effort is required per transaction.

The sender on the other hand needs little-to-no setup (apart from using a supporting wallet) and can achieve an identical or near-identical user experience to a regular transaction, depending on their choices and the sophistication of the implementation.

### What is a Payjoin?

{% include picture.html
   image = "assets/images/guide/case-studies/payjoin case study/payjoin-bip78-process.png"
   retina = "assets/images/guide/case-studies/payjoin case study/payjoin-bip78-process@2x.png"
   alt-text = "Diagram of a BIP-78 payjoin process"
   width = 800
   height = 683
%}

A bitcoin transaction where the sender and receiver construct a transaction together by contributing inputs is called a Payjoin.

It is also called pay-to-endpoint (P2EP) since the coordination is mediated through an endpoint run by the receiver. The endpoint address is communicated through a BIP-21 URI along with the payment address and amount.

> In the payjoin process, parties edit, sign and pass iterations of the transaction between each other before a final version is broadcasted. [BIP-78](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#user-content-Respecting_the_minimum_relay_fee_policy) codifies a protocol with 2 iterations, as shown in the visual here.

Reviewer 2: ```suggestion
In the payjoin process, parties edit, sign and pass iterations of the transaction between each other before a final version is broadcasted. [BIP-78](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#user-content-Respecting_the_minimum_relay_fee_policy) codifies a protocol with one round of interaction beyond address sharing, as shown in the visual here.
```

In studying interactive protocols, I've always seen and have come to expect interaction quantified in "rounds" from round-robin tournament terminology.

Author: Someone else (or you) made the point in the google doc few days back. In an extended payjoin scenario where the sender edits/signs and sends it back to receiver to sign/broadcast the tx, will it be considered 1.5 rounds?

Author: I'm open to adding this as an alternative phrasing...

{% include picture.html
   image = "assets/images/guide/case-studies/payjoin case study/payjoin sender flow.png"
   retina = "assets/images/guide/case-studies/payjoin case study/payjoin sender flow@2x.png"
   alt-text = "payjoin sender flow"
   width = 800
   height = 374
%}

The detailed thought process can be read [here](https://docs.google.com/document/d/1IPCQsdoVBGEceUnWhOeEhi-xcNF049RG1tRAchz-Iuc/edit?usp=sharing).

View this flow in [Figjam here](https://www.figma.com/file/69uUDWVc8N9t5Bej8pZEsF/PayJoin-User-Flows?node-id=0%3A1&t=8eOBQa7JofjZ3p3z-1). The detailed process and considerations can be reviewed in this [FigJam file](https://www.figma.com/file/NzMvwyzP7x5jfGmwNUKRov/PayJoin-Process-Flows?node-id=336%3A619&t=BhGymSWCzl2Xm2oO-1).

##### The Payjoin handshake

<p class="h3">The term ‘Payjoin handshake’ is being introduced to succinctly describe the process of sending a original PSBT to receiver’s endpoint, receiver contributing their inputs, modifying the transaction as needed and signing the new draft of the transaction.</p>

> The above steps are not important to the users involved in the transaction, and do not need specific user action from either the sender or receiver. More generally, the transacting parties need not even be aware that this is occurring.

Reviewer 2: I  love to see innovation in the way we can explain the interaction! Way easier to understand what's going on if we can  put consise words to it. However, I think "payjoin *rendezvous*" is better suited than "payjoin handshake" in this context.

In the context of an interactive protocol (like TLS), a handshake refers to a series of steps or messages exchanged between two parties to establish a connection and initialize communication. In payjoin, the communication is already over after the steps described here as a handshake are complete

A rendezvous refers to a mechanism or procedure for two parties to synchronize and agree on the parameters of an interaction.

Author: I can see the confusion for a technical reader but the 'handshake' is intended for a colloquial, non-technical purpose/audience. It's also clearer & simpler (to me) than the 'rendezvous'.

Reviewer 2: I'm not even sold on rendezvous. It would be good to say *why* the term needs to be introduced. Did you run into confusion in existing UI where there was no explanation of what was going on? And introducing new vocabulary creates a better experience?

Author: I think so, addressed this in the case study text.

   width = 800
   height = 754
%}

A detailed explainer of each screen can be viewed [here](https://docs.google.com/document/d/1laHP3TqU6zzf0ajpQfQ5qaP0wUUQXNNnoMvVrVVU7K0/edit?usp=sharing). The figma file with the designs is [here](https://www.figma.com/file/hCHA4qjxQGiX06ddnWADyW/%5Byashraj's-copy%5D-Bitcoin-Wallet-UI-Kit-%26-Design-System?node-id=4331%3A66395&t=uV0lFmRXrQiv2AAB-1).

### Receiving Payjoins

Based on the BIP-78 protocol, [the receiver has higher requirements](http://0.0.0.0:4000/guide/case-studies/payjoin-case-study/#understanding-users--requirements) than the sender does. While senders need a compatible mobile wallet, they cannot serve an always-online endpoint. Businesses & institutions on other hand could implement a payjoin receiver setup easily, but avoid hot wallets due to security concerns.

### While practically every mobile wallet is a hot wallet and it is trivial to fund it, having an online server where the payjoin handshake can be performed is difficult for technical and practical reasons. This might be the biggest impediment to payjoin support and adoption so far.

#### Investigation: BTCPay merchant/receiving

> The author successfully tested receiving payjoins (BTCPay 1.7.12) in Feb 2023.

Reviewer 2: ```suggestion
The author successfully tested receiving payjoins (BTCPay v1.7.12) in Feb 2023.
```

---
layout: guide
title: The Payjoin experience
description: This case study is an analysis and exploration of payjoin's audience, implementations, features and proposes specific user flows.
nav_order: 6
parent: Case studies
permalink: /guide/case-studies/payjoin/
main_classes: -no-top-padding

> image: https://bitcoin.design/assets/images/guide/case-studies/blixt-wallet/blixt-wallet-preview.jpg

Author: ```suggestion
image: "assets/images/guide/case-studies/payjoin-case-study/header-graphic-metaphor-preview-image.png"
```

Reviewer: You probably don't want the `"` at the start and end of the file name

    - file: send-payjoin-success
      modalImage: send-payjoin-success-big
      alt: A payment success screen informs the user that the Payjoin transaction has been broadcast.
      caption: A payment success screen informs the user that the Payjoin transaction has been broadcast.

---

<!--

Editor's notes

Illustration sources: <>

-->

{% include picture.html
   image = "assets/images/guide/case-studies/payjoin/header-graphic-metaphor-banner.png"
   retina = "assets/images/guide/case-studies/payjoin/header-graphic-metaphor-banner@2x.png"
   alt-text = "Payjoin case study banner image"
   width = 1600
   height = 600
   layout = "full-width"
%}

# The Payjoin Experience
{:.no_toc}
---

<div class="glossary-toc" markdown="1">
* Table of contents
{:toc}
</div>

---

A bitcoin transaction where the sender and receiver collaboratively construct a transaction is called a Payjoin. Payjoin transactions can be a powerful tool providing privacy, fee-savings and UTXO consolidation benefits to users. They also present challenges and trade-offs due to their interactive & synchronous nature.

> **Payjoin transactions break some of the chain analysis [heuristics](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#motivation) during the course of making a payment. However, real time coordination between the sender & receiver mediated through an endpoint is required to construct them.**

Author: "Real time" is valid usage. Accepting suggestion which is valid as well.

This case study:

- Outlines the incentives and objectives for both senders & receivers

> - Analyses the Payjoin UX on both sides and proposes approaches to design Payjoin experiences.

Reviewer: ```suggestion
- Analyses the payjoin UX on both sides and proposes approaches to design payjoin experiences.
```

- Proposes specific user flows and where possible, working designs for the same.

The high-level findings from this case study are that while the setup is more involved for the receiver, little-to-no effort is required per transaction.

The sender on the other hand needs little-to-no setup (apart from using a supporting wallet) and can achieve an identical or near-identical user experience to a regular transaction, depending on their choices and the sophistication of the implementation.

> ### What is a Payjoin?

Reviewer: ```suggestion
### What is a payjoin?
```

> In a payjoin, both the sender and receiver contribute inputs to the transaction in a coordinated manner. The payjoin mechanism is also called pay-to-endpoint (P2EP) since the coordination is mediated through an endpoint run by the receiver. The endpoint address is communicated through a [BIP-21](https://bitcoinqr.dev) URI along with the payment address and amount.

Reviewer: ```suggestion
In a payjoin, both the sender and receiver contribute inputs to the transaction in a coordinated manner. The payjoin mechanism is also called pay-to-endpoint (P2EP), since the coordination is mediated through an endpoint run by the receiver. The endpoint address is communicated through a [BIP-21](https://bitcoinqr.dev) URI, along with the payment address and amount.
```

> In the payjoin process, parties edit, sign and pass iterations of the transaction between each other before a final version is broadcasted. [BIP-78](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#user-content-Respecting_the_minimum_relay_fee_policy) codifies a protocol with 2 iterations (or one round of interaction beyond address sharing), as shown in the visual here.

Reviewer: ```suggestion
In the payjoin process, parties edit, sign and pass iterations of the transaction between each other, before a final version is broadcasted. [BIP-78](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#user-content-Respecting_the_minimum_relay_fee_policy) codifies a protocol with 2 iterations (or one round of interaction beyond address sharing), as shown in the visual here.
```

{% include picture.html
   image = "assets/images/guide/case-studies/payjoin/payjoin-bip78-process.png"
   retina = "assets/images/guide/case-studies/payjoin/payjoin-bip78-process@2x.png"
   alt-text = "Diagram of a BIP-78 payjoin process"
   width = 800
   height = 683
%}

### Background

This case study has its origins in a [design challenge]({{ '/guide/resources/design-challenges/#challenge-6-private-purchase' | relative_url }}), which entailed creating a sender flow for payment using a Payjoin transaction.

While payjoin provides privacy improvements and potential fee-savings, they present the challenge of coordination between transacting parties and may require new user flows to implement features enabled through [BIP-78](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki).

At this point in early 2023, payjoins have not achieved significant adoption in the bitcoin ecosystem.

## Our Approach

This case study followed these steps:

- Understanding users & technical requirements for senders & receivers

> - Exploring the payjoin implementations: BlueWallet (sending) and BTCPay Server (receiving)

Reviewer: ```suggestion
- Exploring the payjoin implementations: [BlueWallet](https://bluewallet.io) (sending) and [BTCPay Server](https://btcpayserver.org) (receiving)
```

- Creating user flows for senders & receivers based on the above

### Understanding users & requirements

> Technically, everyone conducting transactions on the Bitcoin blockchain can use payjoin. This case study categorises users in the following ways:

Reviewer: ```suggestion
Technically, everyone conducting transactions on the bitcoin blockchain can use payjoin. This case study categorises users in the following ways:
```

1. Institutions (businesses, exchanges etc.) and individuals

> 2. sender and receiver of a payment

Reviewer: ```suggestion
2. Sender and receiver of a payment
```

   alt-text = "Payjoin users, goals and stories"
   width = 800
   height = 939
%}

<div markdown="1">
   <table>
   <thead>
      <tr>
         <th>
            Party
         </th>
         <th>
            Requirements
         </th>
         <th>
            Benefits
         </th>
      </tr>
   </thead>
   <tbody>
      <tr>
         <td>Sender</td>
         <td>
            <ul>
               <li>Wallet support</li>
               <li>Internet connection</li>
            </ul>
         </td>
         <td>
            <ul>
               <li>Improved Privacy</li>
            </ul>
         </td>
      </tr>
      <tr>
         <td>Receiver</td>
         <td>
            <ul>
               <li>Hot wallet (for signing transactions)</li>
               <li>Serve/maintain secure web endpoint</li>
               <li>Wallet support</li>
               <li>Funds in hot wallet</li>
            </ul>
         </td>
         <td>
            <ul>
               <li>Privacy</li>
               <li>UTXO consolidation</li>
               <li>Fee savings</li>
            </ul>
         </td>
      </tr>
   </tbody>
</table>
</div>

> ### Sending Payjoins

Reviewer: ```suggestion
### Sending payjoins
```

Participating in payjoin transactions is relatively straightforward for a sender. They only need a supporting wallet and have to be online during the process.

#### Investigation: BlueWallet Breakdown

We analyzed the payjoin user flow in BlueWallet's pioneering implementation from 2020. Read the [detailed investigation](https://docs.google.com/document/d/1_de2pkMREGpZQwOefQdH-MZNSUPazObgRK1cruiWBmo/edit#bookmark=id.jamlvi37p922).

{% include picture.html
   image = "assets/images/guide/case-studies/payjoin/bluewallet-sender-flow.png"
   retina = "assets/images/guide/case-studies/payjoin/bluewallet-sender-flow@2x.png"
   mobile = "assets/images/guide/case-studies/payjoin/bluewallet-sender-flow-mobile.png"
   mobileRetina = "assets/images/guide/case-studies/payjoin/bluewallet-sender-flow-mobile@2x.png"
   alt-text = "bluewallet sender flow"
   width = 800
   height = 54
%}

Here are the findings:

- The fee rate drops once transaction size increases during the payjoin process, which may change estimated confirmation time
- The flow results in a transaction that has a round fee-rate after removing one input, which may be used by chain analysis firms to spot payjoins

> #### The Payjoin Sender Flow

Reviewer: ```suggestion
#### The payjoin sender flow
```

   height = 374
%}

The detailed thought process can be read [here](https://docs.google.com/document/d/1IPCQsdoVBGEceUnWhOeEhi-xcNF049RG1tRAchz-Iuc/edit?usp=sharing).

View this flow in [Figjam here](https://www.figma.com/file/69uUDWVc8N9t5Bej8pZEsF/PayJoin-User-Flows?node-id=0%3A1&t=8eOBQa7JofjZ3p3z-1). The detailed process and considerations can be reviewed in this [FigJam file](https://www.figma.com/file/NzMvwyzP7x5jfGmwNUKRov/PayJoin-Process-Flows?node-id=336%3A619&t=BhGymSWCzl2Xm2oO-1).

##### The Payjoin handshake

**This term is being introduced to succinctly describe the automated back-and-forth process between the sender and the receiver – where the original [PSBT]({{ '/guide/glossary/#partially-signed-bitcoin-transaction-psbt' | relative_url }}) containing only the sender's inputs is transformed into a signed, broadcastable transaction containing inputs from both parties.**

The above steps are not important to the users involved in the transaction, and do not require specific user action from either the sender or receiver. More generally, the transacting parties need not even be aware that this is occurring, but must ensure they remain online during the process.

> ##### Prototypes for the Payjoin sender flow

Reviewer: ```suggestion
##### Prototypes for the payjoin sender flow
```

> **The intention is to keep the user-flow as close to the default flow as possible. Where it becomes necessary to deviate, we educate and help the user to navigate the UI.**

Reviewer: ```suggestion
**The intention is to keep the user flow as close to the default flow as possible. Where it becomes necessary to deviate, we educate and help the user to navigate the UI.**
```

{% include image-gallery.html pages = page.sender-flow-screens %}

A detailed explainer of each screen can be viewed [here](https://docs.google.com/document/d/1laHP3TqU6zzf0ajpQfQ5qaP0wUUQXNNnoMvVrVVU7K0/edit?usp=sharing). The figma file with the designs is [here](https://www.figma.com/file/hCHA4qjxQGiX06ddnWADyW/%5Byashraj's-copy%5D-Bitcoin-Wallet-UI-Kit-%26-Design-System?node-id=4331%3A66395&t=uV0lFmRXrQiv2AAB-1).

> ### Receiving Payjoins

Reviewer: ```suggestion
### Receiving payjoins
```

Here are the insights:

- Generating a payjoin-enabled payment link is automatic once it is setup
- If the user chooses to add an existing hot wallet, BTCPay does not provide an option to enable payjoin on the same page like it does during new hot wallet creation
- If a user opts for a watch-only (cold) wallet during setup (new or existing), they will never encounter payjoin (because a wallet is required in order to sign the payjoin transactions)
- Only one bitcoin wallet can be connected to the store (except through [BTCPay Vault](https://docs.btcpayserver.org/HardwareWalletIntegration/)) so using a cold wallet precludes setting up a hot wallet (and Payjoin)
- BTCPay does not prompt or warn the user if there are no funds in the hot wallet to perform a payjoin, resulting in failed payments
- There is no dedicated space to manage and monitor Payjoin

Read the full analysis on [receiving payjoin payments on BTCPay](https://docs.google.com/document/d/1_de2pkMREGpZQwOefQdH-MZNSUPazObgRK1cruiWBmo/edit?usp=sharing) here.

#### Proposed Receiver Flow

Here we will devise user flows for a point-of-sale (POS) system that can always be online. This excludes mobile wallets.

> A receiver should be able to set up payjoin during onboarding on the platform (app or POS system) or any time after that. The following is a standalone payjoin setup flow.

Reviewer: ```suggestion
A receiver should be able to set up payjoins during onboarding on the platform (app or POS system), or any time after that. The following is a standalone payjoin setup flow.
```

- Payjoins are a great way to consolidate UTXOs so participants pay lower fees in the long run. This can be especially useful for trading desks at exchanges who need this the most, and possess the technical ability to implement a robust setup
- Due to the back-&-forth between parties which can extend over multiple iterations/rounds, this case study introduces the term ‘payjoin handshake’ to encompass all the events between user actions
- Maintaining an always-online endpoint seems to be the biggest hurdle for payjoin implementation, we list some alternatives here
- Payjoin burdens the receiver with more requirements, but also provides more benefits and opportunities
- Address substitution is a powerful idea from BIP-78: it is a risk that can become a powerful ally for the payjoiner
- Instead of setting up another wallet, Payjoin receivers can leverage their daily spending wallet, which is used by the receiver as the Payjoin hot wallet
- Payjoin implementations should be mindful of the round-fee-rate heuristic for identifying payjoins, handle it without need for user decisions

> ## New Horizons

Reviewer: ```suggestion
## New horizons
```

Payjoins can be complicated to understand or implement given the multiple parties and moving parts (such as the endpoint) involved, but this also opens the door for multiple features and services.

Here are some examples:

- Third-party endpoints could be used by receivers (turning off address substitution), trading some privacy for convenience
- There is exciting work underway around [serverless](https://gist.github.com/DanGould/243e418752fff760c9f6b23bba8a32f9) payjoin implementation that would work great on platforms implementing both bitcoin & lightning
- Receiver-side implementations could use address substitution to let users make payments of their own
- Payjoin coordination over NFC (or other wireless near-range communication technology)

> - BIP-78 could to allow 1 more iteration where the sender can adjust fees post-handshake & send it back to the receiver to broadcast

Reviewer: ```suggestion
- BIP-78 could be used to allow 1 more iteration where the sender can adjust fees post-handshake & send it back to the receiver to broadcast
```

This case study tackles the design aspects around payjoin and it hopes to boost interest in it's various use cases and benefits.

*Gratitude:*

> - *[Pavlenex](https://pavlenex.com/) from BTCPay for providing access to a Payjoin-capable btcpay instance to test receiving*

Reviewer: ```suggestion
- *[Pavlenex](https://pavlenex.com/) from BTCPay for providing access to a payjoin-capable BTCPay instance to test receiving*
```

---
layout: guide
title: The Payjoin Experience
description: <>
nav_order: 5
parent: Case studies

> permalink: /guide/case-studies/payjoin-case-study/

Reviewer: Could you please move this page to `/guide/case-studies/payjoin/`? No need to repeat `case study` twice in the URL. That also means moving files and folders around and updating various URLs in this PR.

redirect_from:
 - /guide/payments/contacts/
main_classes: -no-top-padding

> image: https://bitcoin.design/assets/images/guide/case-studies/blixt-wallet/blixt-wallet-preview.jpg

Author: Need help fixing this.

      alt: The user is presented with options of fee-ranges to choose from, in an interface very close to the default.
      caption: The user is presented with options of fee-ranges to choose from, in an interface very close to the default.
    - file: payjoin-handshake-transition
      modalImage: payjoin-handshake-transition@2x
      alt: A transition screen is displayed indicating the payjoin construction and signing process.
      caption: A transition screen is displayed indicating the payjoin construction and signing process.
    - file: send-review
      modalImage: send-review@2x
      alt: The pre-broadcast review screen with the finalised fee after the payjoin handshake is completed.
      caption: The pre-broadcast review screen with the finalised fee after the payjoin handshake is completed.
    - file: send-payjoin-success
      modalImage: send-payjoin-success@2x
      alt: A payment success screen informs the user that the Payjoin transaction has been broadcast.
      caption: A payment success screen informs the user that the Payjoin transaction has been broadcast.

---

<!--

Editor's notes

Illustration sources: <>

-->

<!-----
Conversion notes:

> * Docs to Markdown version 1.0β34

Author: Fixed.

* Sun Mar 05 2023 13:50:07 GMT-0800 (PST)
* Source doc: [restructure] Payments with Payjoin: A Case Study
* Tables are currently converted to HTML tables.
* This document has images: check for >>>>>  gd2md-html alert:  inline image link in generated source and store images to your server. NOTE: Images in exported zip file from Google Docs may not appear in  the same order as they do in your doc. Please check the images!

----->
{% include picture.html
   image = "assets/images/guide/case-studies/payjoin-case-study/header-graphic-metaphor-banner.png"
   retina = "assets/images/guide/case-studies/payjoin-case-study/header-graphic-metaphor-banner@2x.png"
   alt-text = "Payjoin case study banner image"
   width = 1600
   height = 600
   layout = "full-width"
%}

# The Payjoin Experience
{:.no_toc}

---

<div class="glossary-toc" markdown="1">
* Table of contents
{:toc}
</div>

---

> ## Introduction

Author: Great suggestion.

At this point in early 2023, payjoins have not achieved significant adoption in the bitcoin ecosystem.

## Our Approach

This case study follows these tentative steps:

- Understanding users & technical requirements for senders & receivers
- Exploring the payjoin implementations: BlueWallet (sending) and BTCPay Server (receiving)
- Creating user flows for senders & receivers based on the above

### Understanding users & requirements

Technically, everyone conducting transactions on the Bitcoin blockchain can use payjoin. This case study categorises users in the following ways:

1. Institutions (businesses, bitcoin exchanges etc.) and individuals
2. sender and receiver of a payment

{% include picture.html
   image = "assets/images/guide/case-studies/payjoin-case-study/users, goals and stories.png"
   retina = "assets/images/guide/case-studies/payjoin-case-study/users, goals and stories@2x.png"
   alt-text = "Payjoin users, goals and stories"
   width = 800
   height = 939
%}

> <div markdown="1">

Author: Not sure how to address this.

            </ul>
         </td>
      </tr>
   </tbody>
</table>
</div>

### Sending Payjoins

Participating in payjoin transactions is relatively straightforward for a sender. They only need a supporting wallet and have to be online during the process.

#### Investigation: Blue Wallet [Video](https://www.youtube.com/watch?v=xowwF2yFSZ8) Breakdown

We analyzed the payjoin user flow in BlueWallet. Read the detailed investigation [here](https://docs.google.com/document/d/1_de2pkMREGpZQwOefQdH-MZNSUPazObgRK1cruiWBmo/edit#bookmark=id.jamlvi37p922).

{% include picture.html
   image = "assets/images/guide/case-studies/payjoin-case-study/bluewallet-sender-flow.png"
   retina = "assets/images/guide/case-studies/payjoin-case-study/bluewallet-sender-flow@2x.png"
   mobile = "assets/images/guide/case-studies/payjoin-case-study/bluewallet-sender-flow-mobile.png"
   mobileRetina = "assets/images/guide/case-studies/payjoin-case-study/bluewallet-sender-flow-mobile@2x.png"
   alt-text = "bluewallet sender flow"
   width = 800
   height = 54
%}

Here are the findings:

- The fee rate drops once transaction size increases during the payjoin process,  which may change estimated confirmation time

> - The flow results in a transaction that has a round amount fee-rate after removing one input, which may be used by chain analysis firms to spot payjoins

Reviewer 2: See "Round payment amount" used to distinguish between change / payment [here](https://blockstream.info/tx/b2ddf68bf2318b51ec5e6bdc7831e7063cd04c0ff84edbb56808de62fbb0de74)

I believe the case study hightlights how a similar round fee-rate heuristic may distinguish the payment amount from change here.

https://en.bitcoin.it/wiki/Privacy#Round_numbers

> - The flow results in a transaction that has a round amount fee-rate after removing one input, which may be used by chain analysis firms to spot payjoins

Author: Thank you, do you suggest any changes/clarifications to my wording?

#### The Payjoin Sender Flow

We designed a sender flow that attempts to address the issues identified above.

For the purpose of creating the sender flow, we will assume that the receiver only contributes UTXOs but does not contribute to the fees since this allows a simple, automated process for both parties.

> ### In short, the sender flow outlined here asks the user to choose a fee-range instead of a fee-amount (or fee-rate) while keeping the rest of the user flow almost exactly the same. We use it to set 3 optional parameters specified in BIP-78, which can be used to construct a simple but effective payjoin implementation.

Reviewer: ```suggestion
** In short, the sender flow outlined here asks the user to choose a fee-range instead of a fee-amount (or fee-rate) while keeping the rest of the user flow almost exactly the same. We use it to set 3 optional parameters specified in BIP-78, which can be used to construct a simple but effective payjoin implementation.**
```

{% include picture.html
   image = "assets/images/guide/case-studies/payjoin-case-study/payjoin-sender-flow.png"
   retina = "assets/images/guide/case-studies/payjoin-case-study/payjoin-sender-flow@2x.png"
   alt-text = "payjoin sender flow"
   width = 800
   height = 374
%}

The detailed thought process can be read [here](https://docs.google.com/document/d/1IPCQsdoVBGEceUnWhOeEhi-xcNF049RG1tRAchz-Iuc/edit?usp=sharing).

View this flow in [Figjam here](https://www.figma.com/file/69uUDWVc8N9t5Bej8pZEsF/PayJoin-User-Flows?node-id=0%3A1&t=8eOBQa7JofjZ3p3z-1). The detailed process and considerations can be reviewed in this [FigJam file](https://www.figma.com/file/NzMvwyzP7x5jfGmwNUKRov/PayJoin-Process-Flows?node-id=336%3A619&t=BhGymSWCzl2Xm2oO-1).

##### The Payjoin handshake

> ### The term ‘Payjoin handshake’ is being introduced to succinctly describe the process of sending a original PSBT to receiver’s endpoint, receiver contributing their inputs, modifying the transaction as needed and signing the new draft of the transaction.</p>

Reviewer: ```suggestion
** The term ‘Payjoin handshake’ is being introduced to succinctly describe the process of sending a original PSBT to receiver’s endpoint, receiver contributing their inputs, modifying the transaction as needed and signing the new draft of the transaction.**
```

The above steps are not important to the users involved in the transaction, and do not need specific user action from either the sender or receiver. More generally, the transacting parties need not even be aware that this is occurring.

##### Prototypes for the Payjoin sender flow

> ### The intention is to keep the user-flow as close to the default flow as possible. Where it becomes necessary to deviate, we educate and help the user to navigate the UI.

Reviewer: ```suggestion
**The intention is to keep the user-flow as close to the default flow as possible. Where it becomes necessary to deviate, we educate and help the user to navigate the UI.**
```

> <div class="image-slide-gallery">

Reviewer: Don't need this line, the `image-gallery.html` include already has it.

> <div class="image-slide-gallery">

Author: ```suggestion

```

{% include image-gallery.html pages = page.sender-flow-screens %}

> </div>

Reviewer: Don't need this line, the `image-gallery.html` include already has it.

A detailed explainer of each screen can be viewed [here](https://docs.google.com/document/d/1laHP3TqU6zzf0ajpQfQ5qaP0wUUQXNNnoMvVrVVU7K0/edit?usp=sharing). The figma file with the designs is [here](https://www.figma.com/file/hCHA4qjxQGiX06ddnWADyW/%5Byashraj's-copy%5D-Bitcoin-Wallet-UI-Kit-%26-Design-System?node-id=4331%3A66395&t=uV0lFmRXrQiv2AAB-1).

### Receiving Payjoins

> Based on the BIP-78 protocol, [the receiver has higher requirements](http://0.0.0.0:4000/guide/case-studies/payjoin-case-study/#understanding-users--requirements) than the sender does. While senders need a compatible mobile wallet, they cannot serve an always-online endpoint. Businesses & institutions on other hand could implement a payjoin receiver setup easily, but avoid hot wallets due to security concerns.

Reviewer: ```suggestion
Based on the BIP-78 protocol, [the receiver has higher requirements]({{ '/guide/case-studies/payjoin-case-study/#understanding-users--requirements' | relative_url }}) than the sender does. While senders need a compatible mobile wallet, they cannot serve an always-online endpoint. Businesses & institutions on other hand could implement a payjoin receiver setup easily, but avoid hot wallets due to security concerns.
```
Don't want to link to your local dev environment here.

> ### While practically every mobile wallet is a hot wallet and it is trivial to fund it, having an online server where the payjoin handshake can be performed is difficult for technical and practical reasons. This might be the biggest impediment to payjoin support and adoption so far.

Reviewer: ```suggestion
**While practically every mobile wallet is a hot wallet and it is trivial to fund it, having an online server where the payjoin handshake can be performed is difficult for technical and practical reasons. This might be the biggest impediment to payjoin support and adoption so far.**
```

#### Investigation: BTCPay merchant/receiving

The author successfully tested receiving payjoins (BTCPay v1.7.12) in Feb 2023.

Here are problems identified:

> - Generating a payjoin-enabled payment link is automatic once it is setup

Reviewer: Why is it a problem that it's automatic?

Author: Fixed.

- If the user chooses to add an existing hot wallet, BTCPay does not provide an option to enable payjoin on the same page like it does during new hot wallet creation
- If a user opts for a watch-only (cold) wallet during setup (new or existing), they will never encounter payjoin (because a wallet is required in order to sign the payjoin transactions)
- Only one bitcoin wallet can be connected to the store (except through [BTCPay Vault](https://docs.btcpayserver.org/HardwareWalletIntegration/)) so using a cold wallet precludes setting up a hot wallet (and Payjoin)
- BTCPay does not prompt or warn the user if there are no funds in the hot wallet to perform a payjoin, resulting in failed payments
- There is no dedicated space to manage and monitor Payjoin

Read the full analysis on [receiving payjoin payments on BTCPay](https://docs.google.com/document/d/1_de2pkMREGpZQwOefQdH-MZNSUPazObgRK1cruiWBmo/edit?usp=sharing) here.

#### Proposed Receiver Flow

> Here we will devise user flows for a POS system that can always be online. This excludes mobile wallets.

Reviewer: ```suggestion
Here we will devise user flows for a point-of-sale (POS) system that can always be online. This excludes mobile wallets.
```
Make sure to explain abbreviations when you introduce them.

A receiver should be able to set up payjoin during onboarding on the platform (app or POS system) or any time after that. The following is a standalone payjoin setup flow.

{% include picture.html

>    image = "assets/images/guide/case-studies/payjoin-case-study/payjoin-receiver-flow@2x.png"

Reviewer: ```suggestion
   image = "assets/images/guide/case-studies/payjoin-case-study/payjoin-receiver-flow.png"
```

   retina = "assets/images/guide/case-studies/payjoin-case-study/payjoin-receiver-flow@2x.png"
   alt-text = "payjoin receiver flow"
   width = 800
   height = 458
%}

A detailed process flow integrated with onboarding is available in this [Figjam file](https://www.figma.com/file/NzMvwyzP7x5jfGmwNUKRov/PayJoin-Process-Flows?node-id=0%3A1&t=wSwewTOkQddWhagl-1).

##### Requesting payjoin

Once the payjoin setup is functional, all payment requests (BIP-21 payment links or encoded as QRs) should contain the necessary information (including endpoint address) to perform the payjoin transaction. No separate user action should be required.

#####  Payjoin Status/Settings

A dedicated section in the settings page might be a good idea to provide dedicated space in the UI to manage Payjoin settings and monitor its status. This would also provide UI space for users using a cold wallet to discover and set up payjoin. This section could contain:

- Payjoin setup workflow
- Additional payjoin features like third party endpoints, address substitution, cold-wallet backfill

> - Monitor payjoin readiness: hot wallet status, pj-endpoint status etc.

Reviewer: ```suggestion
- Monitor payjoin readiness: hot wallet status, Payjoin-endpoint status etc.
```

## Insights

We learnt many things over the course of this case study, and the user flows we have designed already use many of these insights.

> - Payjoins are a great way to consolidate UTXOs so they pay lower fees in the long run – this can be especially useful to exchanges, trading desks who need this the most and possess technical ability to implement a robust setup

Reviewer: ```suggestion
- Payjoins are a great way to consolidate UTXOs so participants pay lower fees in the long run. This can be especially useful for trading desks at exchanges who need this the most, and possess the technical ability to implement a robust setup
```

> - Due to the back-&-forth between parties which can extend over multiple iterations/rounds, this case study introduces the term ‘payjoin handshake’ to encompass all the events between user actions

Reviewer: ```suggestion
- Due to the back-and-forth between parties which can extend over multiple iterations/rounds, this case study introduces the term ‘Payjoin handshake’ to encompass all the events between user actions
```

- Maintaining an always-online endpoint seems to be the biggest hurdle for payjoin implementation, we list some alternatives here

> - Payjoin burdens the receiver with more requirements but also provides more benefits and opportunities

Reviewer: ```suggestion
- Payjoin burdens the receiver with more requirements, but also provides more benefits and opportunities
```

- Address substitution is a powerful idea from BIP-78: it is a risk that can become a powerful ally for the payjoiner

> - Payjoins receivers can leverage their daily spending wallet-saving wallet setup, where the daily spending is used by the receiver as the payjoin hot wallet

Reviewer: ```suggestion
- Instead of setting up another wallet, Payjoin receivers can leverage their daily spending wallet, which is used by the receiver as the Payjoin hot wallet
```

> - Payjoin implementations should be mindful of the round-fee-rate heuristic for identifying payjoins, handle it without need for user decisions

Reviewer: ```suggestion
- Payjoin implementations should be mindful of the round-fee-rate heuristic for identifying Payjoins, and handle it without a need for complex user decisions
```

## New Horizons

> Payjoins can be complicated to understand or implement given the multiple parties and moving parts (such as the endpoint) involved, but this also opens the door for multiple features and services.

Reviewer: ```suggestion
Payjoins can be complicated to understand or implement, given the multiple parties and moving parts (such as the endpoint) involved. But, this also opens the door for multiple features and services.
```

Author: Agree it is kinda long as a bullet point, but I avoid using periods in bullet points.

Here are some examples:

> - Third party endpoints could be used by receivers (turning-off address substitution), trading some privacy for convenience

Reviewer: ```suggestion
- Third-party endpoints could be used by receivers (turning off address substitution), trading some privacy for convenience
```

> - There is exciting work underway around [serverless](https://gist.github.com/DanGould/243e418752fff760c9f6b23bba8a32f9) payjoin implementation that would work great on platforms with both bitcoin & lightning

Reviewer: ```suggestion
- There is exciting work under way around [serverless](https://gist.github.com/DanGould/243e418752fff760c9f6b23bba8a32f9) Payjoin implementations that would work great on platforms implementing both bitcoin & lightning
```

Author: @sbddesign and I settled on using payjoin like we use bitcoin, will incorporate the other suggestion.

Reviewer: Great. Still seeing lots of capitalized used of payjoin in sentences throughout the page, so I'll wait with reviewing.

- Receiver-side implementations could use address substitution to let users make payments of their own
- Payjoin coordination over NFC (or other wireless near-range communication technology)

> - BIP-78 could to allow 1 more iteration where sender can adjust fees post-handshake & send it back to receiver to broadcast

Reviewer: ```suggestion
- BIP-78 could to allow 1 more iteration where senders can adjust fees post-handshake & send the transaction back to receivers to broadcast
```

### Background

The payjoin case study has its origins in a [design challenge]({{ '/guide/resources/design-challenges/#challenge-6-private-purchase' | relative_url }}), which entailed creating a sender flow for payment using a Payjoin transaction.

> While payjoin provides privacy improvements and potential fee-savings, they present the challenge of coordination between transacting parties and may require new user flows to implement features enabled through [BIP-78](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki).

Reviewer 3: ```suggestion
While payjoin provides privacy improvements and potential fee-savings, they present  unique challenges. One of these is the need of coordination between transacting parties to implement features through [BIP-78](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki). This would require modifications to the userflow. 
```

##### The Payjoin handshake

> ** The term ‘Payjoin handshake’ is being introduced to succinctly describe the process of sending a original PSBT to receiver’s endpoint, receiver contributing their inputs, modifying the transaction as needed and signing the new draft of the transaction.**

Reviewer 3: ```suggestion
** The term ‘Payjoin handshake’ is being introduced to succinctly describe the process of sending a original PSBT to receiver’s endpoint, receiver contributing their inputs, modifying the transaction as needed and sign the new draft of the transaction.**
```

> The above steps are not important to the users involved in the transaction, and do not need specific user action from either the sender or receiver. More generally, the transacting parties need not even be aware that this is occurring.

Reviewer 3: ```suggestion
The above steps are not important to the users involved in the transaction, and do not need specific user action from either the sender or receiver. More generally, the transacting parties do not even be aware that this is occurring.
```

### Receiving Payjoins

> Based on the BIP-78 protocol, [the receiver has higher requirements]({{ '/guide/case-studies/payjoin-case-study/#understanding-users--requirements' | relative_url }}) than the sender does. While senders need a compatible mobile wallet, they cannot serve an always-online endpoint. Businesses & institutions on other hand could implement a payjoin receiver setup easily, but avoid hot wallets due to security concerns.

Reviewer 3: ```suggestion
Based on the BIP-78 protocol, [the receiver has higher requirements]({{ '/guide/case-studies/payjoin-case-study/#understanding-users--requirements' | relative_url }}) than the sender does. While senders need a compatible mobile wallet, they cannot serve an always-online endpoint. Businesses & institutions could implement a payjoin receiver setup easily, but avoid hot wallets due to security concerns.
```
